Search ClassifiedListings frontend code

atsmith813 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - Read the DEV Contributing Guide and the Code of Conduct
     - Provided tests for your changes
     - Used descriptive commit messages
     - Updated any relevant documentation and added any necessary screenshots
-->

## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This is the frontend code for moving the search of `ClassifiedListings` over to Elasticsearch.

~**PLEASE NOTE:** we need to update params in Fastly for `/search/classified_listings`, but I don't have access or know how to do that. I believe the params that need to be added right now are `classified_listing_search`, `category`, and `tags`...assuming `page` and `per_page` are already in Fastly? Not sure if that's by endpoint or not.~

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/6#card-33499060

## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

## Note
~Before merging, I'll need some assistance whitelisting the new params in Fastly - I do not have access.~

![red_ranger_running_through_explosions_gif](https://media.giphy.com/media/OOZLyBA9Euq2I/giphy.gif)

atsmith813 commented on app/javascript/listings/listings.jsx:
```diff
@@ -83,10 +77,18 @@ export class Listings extends Component {
+      {
+        leading: true,
+      },
+    );
```
Debouncing, in general, will eventually be refactored out into a more reusable method https://github.com/thepracticaldev/dev.to/pull/6358#discussion_r385807411

vaidehijoshi replied:
IT'S HAPPENING!

vaidehijoshi replied:
Although this seems formatted in a really weird way to me? 🤔 did your editor do this?

atsmith813 replied:
I did it 🙆‍♂️. I like the it reads and it makes future code diffs very easy if we add more options to debounce (unlikely).

atsmith813 commented:
What are everyone's thoughts on codeclimate? I'm not sure what the best approach is.

My 2 cents...I'm fine overriding it and saving the refactor for a separate PR. I imagine it would be better to wait for the last model or two to be brought into Elasticsearch as well.

vaidehijoshi commented:
> My 2 cents...I'm fine overriding it and saving the refactor for a separate PR. I imagine it would be better to wait for the last model or two to be brought into Elasticsearch as well.

Just to play devil's advocate here, if we refactored the repeated blocks of code in actions, listings and tags, we could create a reusable fetch method that could be used for all search-related queries, which could be used in the remaining Elasticsearch PRs 👀 

This isn't a hill I'm willing to die on, but I definitely think we're hitting the point where we _should_ refactor. If that is in the next PR, I'm fine with that too! 👍

atsmith813 commented:
@vaidehijoshi that's all true as well assuming the future models have the same behavior and setup (I'm sure they will).

My hesitation is that if I do that refactor in this PR, it then opens up a whole separate conversation about how that refactor should be done which is outside of the scope of what this PR is going. The review could then become noisy with feedback on 2 separate concepts. If we aren't comfortable overriding codeclimate, I'll close this PR for now and open a new one to do all the refactoring first.

vaidehijoshi commented:
@atsmith813 I think the refactor could be the next PR, if you're okay with going that direction; I definitely don't think you need to close this one to prioritize the refactor :) I could go either way, just as long as we do it before adding more search functions that will repeat this code! I'll defer to what you feel is best and makes the most sense for your workflow here.

atsmith813 commented:
I made a card to refactor the JS as my next task once this is merged https://github.com/thepracticaldev/dev.to/projects/6#card-33848509

atsmith813 commented:
Thanks @mstruve! https://github.com/thepracticaldev/dev.to/pull/6412

atsmith813 commented:
Given we have to wait for #6412 to be merged and deployed first, let's go ahead and get #6374 merged and deployed before this PR as well.

After that, I'll re-base with master, and use the new helper from #6374 to resolve codeclimate here instead of overriding it. At this point, that feels like the cleanest approach!

mstruve commented:
New params, classified_listing_search and tags, have been added to Fastly. category, page, and per_page already exist.

atsmith813 commented:
Re-requesting reviews on this one just to be safe.

I made a couple of updates after I rebased with master to get the updates from #6412 and #6374, hence the force push.
1) I found a bug due to a change made in #6412 that I fixed.
2) I leveraged the helper created in #6374 plus some refactoring to make codeclimate happy so this PR will no longer require an admin override to merge.

atsmith813 commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +301,28 @@ export class Listings extends Component {
+      classified_listing_search: query,
+    };
+
+    if (dataHash.tags) dataHash.tags = tags;
```
After #6412 was merged, we removed some backend checks for empty params. At this stage, `tags` is often an empty string so I made this a conditional to avoid sending `tags=""` to search which ends up returning nothing when if we do let that through.

vaidehijoshi replied:
Since we're not allowing empty param searches for ClassifiedListings specifically, I think we should document that here in the frontend code for this method in particular. If (or when?) we decide to disallow empty param searches for _all_ search endpoints, we can easily figure out which ones have accounted for this or not, and then abstract it into `fetchSearch`.

Another thing I'm curious about: why _not_ just disallow all empty param searches outright? Is there a reason (or benefit) to only make this change to classified listings? Do we ever actually want to allow users to search with empty params and hit the endpoint?

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +301,28 @@ export class Listings extends Component {
+      classified_listing_search: query,
+    };
+
+    if (dataHash.tags) dataHash.tags = tags;
```
The convention in the app so far seems to be this, so I'd prefer to keep it consistent. Again, I feel like we should have eslint enforce these kinds of things, but since it's not at the moment...🤷‍♀ 

```suggestion
    if (dataHash.tags) { dataHash.tags = tags; }
```

atsmith813 replied:
Updated!

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +301,28 @@ export class Listings extends Component {
+
+    if (dataHash.tags) dataHash.tags = tags;
+
+    if (category.length > 0) dataHash.category = category;
```
```suggestion
    if (category.length > 0) { dataHash.category = category; }
```

atsmith813 replied:
Updated!

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +301,28 @@ export class Listings extends Component {
         listings: fullListings,
         initialFetch: false,
-        showNextPageButt: content.hits.length === 75,
+        showNextPageButt: classifiedListings.length === 75,
```
What is `75` here? This feels like a magic number to me. Can we make it a `const` at the top of the file with some documentation for it, and then use the const here?

atsmith813 replied:
Update!

atsmith813 commented:
Hmmm...turns out re-requesting doesn't invalidate a previous approval 🤔 I thought it did. None the less, I still think it's best to get a second pair of eyes on my updates before shipping since this change will affect the UX vs the other PRs on this topic were backend changes that the user didn't interact with yet (until now).

cc @mstruve @rhymes @citizen428

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +301,28 @@ export class Listings extends Component {
-      });
+    const { page, listings } = t.state;
+    const dataHash = {
+      per_page: 75,
```
can reuse a const here too

atsmith813 replied:
Updated!

atsmith813 commented:
@vaidehijoshi my mistake, I worded it poorly.

> Since we're not allowing empty param searches for ClassifiedListings specifically, I think we should document that here in the frontend code for this method in particular. If (or when?) we decide to disallow empty param searches for _all_ search endpoints, we can easily figure out which ones have accounted for this or not, and then abstract it into `fetchSearch`.

> Another thing I'm curious about: why _not_ just disallow all empty param searches outright? Is there a reason (or benefit) to only make this change to classified listings? Do we ever actually want to allow users to search with empty params and hit the endpoint?

We actually do allow empty params. The problem is if you send empty params like `""` it's going to try and filter by that. So in this example, we were basically asking Elasticsearch to return results with `tags` (`tag_list`) containing an empty string, which is none of them.

There could be a situation where we'd want that in the future. If we're searching for records that have a specified field has `null` or maybe something along those lines? I'm not entirely sure either.

This change isn't made only to classified listings. I think it seems that way because the value of `tags` at this point in the code is `""` by default. In other search endpoints, they're not passing in params that default to a value of `""`.

The next step to make this 10/10 perfect would be to change that behavior of tags defaulting to `""`, but that would touch more code outside of this scope so I didn't go that route.

mstruve commented:
> If we're searching for records that have a specified field has null

FWIW in those cases where we want to search for a missing field in ES we will be doing a must not exist search and we would likely have a "none" button of some sort to trigger that.

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -1,6 +1,11 @@
+import { fetchSearch } from '../src/utils/search';
 import SingleListing from './singleListing';
 
+/* How many listings to show per page */
```
shouldn't we use jsdoc style comments here?

atsmith813 replied:
Updated!

vaidehijoshi commented on app/javascript/listings/listings.jsx:
```diff
@@ -284,27 +304,33 @@ export class Listings extends Component {
+      classified_listing_search: query,
+    };
+
+    /* tags defaults to an empty String, so we check for that case here */
```
I think we should use jsdoc to document this whole method, since we're in the code anyways.

atsmith813 replied:
Updated!

vaidehijoshi commented:
> This change isn't made only to classified listings. I think it seems that way because the value of tags at this point in the code is "" by default. In other search endpoints, they're not passing in params that default to a value of "".

I only thought that it was applying to classified listings because I saw [these lines](https://github.com/thepracticaldev/dev.to/pull/6412/files#diff-b94390073d2ab319b2b81c7dcb515293R36-R39) in #6412, the PR you linked to. Where else did this change? Does it impact other places where we hit the `/search` endpoints and pass in params?

> FWIW in those cases where we want to search for a missing field in ES we will be doing a must not exist search and we would likely have a "none" button of some sort to trigger that.

Not entirely sure what this means (because I haven't worked with ES), but based on @mstruve's comment here, it seems like we do want to continue to allow `""` to be passed in in _some_ places?

atsmith813 commented:
> > This change isn't made only to classified listings. I think it seems that way because the value of tags at this point in the code is "" by default. In other search endpoints, they're not passing in params that default to a value of "".
> 
> I only thought that it was applying to classified listings because I saw [these lines](https://github.com/thepracticaldev/dev.to/pull/6412/files#diff-b94390073d2ab319b2b81c7dcb515293R36-R39) in #6412, the PR you linked to. Where else did this change? Does it impact other places where we hit the `/search` endpoints and pass in params?

Ah, I see what you're saying now! My apologies. I'm going to DM you about this one @vaidehijoshi.
> 
> > FWIW in those cases where we want to search for a missing field in ES we will be doing a must not exist search and we would likely have a "none" button of some sort to trigger that.
> 
> Not entirely sure what this means (because I haven't worked with ES), but based on @mstruve's comment here, it seems like we do want to continue to allow `""` to be passed in in _some_ places?

And I think that's correct!

atsmith813 commented on app/javascript/listings/listings.jsx:
```diff
@@ -38,6 +43,20 @@ function resizeAllMasonryItems() {
+
+  classifiedListings.forEach(listing => {
+    if (listing.bumped_at) {
+      if (!listings.map(l => l.id).includes(listing.id)) {
```
I think this was originally done in the event duplicate listings were returned 🤔

I'm not sure that (duplicate listings) are possible under our setup now. Am I thinking of that correctly? Any thoughts on that before I update this?

atsmith813 replied:
Following up, I played around with this. Listings end up being reset to `[]` each time so we don't have to worry about duplicates. I thought maybe there would be a case where the existing `listings` state might include a listing that is also returned from search. But I wasn't able to produce that case. Furthermore, we're not permitting duplicate ClassifiedListings in the backend either so we shouldn't have to worry about duplicates in the response either.

I went ahead and pushed your suggestion, thanks for that!

atsmith813 commented:
@rhymes I'll DM you, that's very interesting and it's not happening to me locally 🤔

Good old, "but it works on my machine" :trollface:

atsmith813 commented on app/javascript/listings/listings.jsx:
```diff
@@ -282,29 +306,38 @@ export class Listings extends Component {
+      page,
+      per_page: LISTING_PAGE_SIZE,
+      tags,
+    };
```
This is much cleaner now. The params cleanup logic has been moved to the `search_controller`.

atsmith813 commented on app/controllers/search_controller.rb:
```diff
@@ -28,8 +45,24 @@ def chat_channel_params
+  # nil differently. This is a helper method to remove any params that are
+  # blank before passing it to Elasticsearch.
+  def remove_blank_params
+    params.delete_if { |_k, v| v.blank? }
```
After a great convo with @vaidehijoshi I've changed this up a little bit. The challenge from a development perspective is that each `QueryBuilder` for each Model in Elasticsearch handles params a little differently. For example, some `QueryBuilders` will clean up `nil` or empty String values automatically. Others might let it through (a behavior we'd like to intentionally keep right now). When making a request from the frontend, a dev basically has to go through a lot to understand what to pass in and how it's treated.

Instead, I'm starting a new trend where we keep general params cleanup here in the `search_controller` with the idea that in the future a dev would only have to check the list of whitelisted params and should be able to get most of the way without having to dig further into the various `Search` modules, including the `QueryBuilder` to understand how things like `""` are being handled.

The controller is also easy to implement clean up across all searches or specific ones vs changing code in _each_ `QueryBuilder` for a similar, all-encompassing change. Eventually, I suspect we'll refactor `QueryBuilders`, but that's further off right now.

vaidehijoshi commented on app/controllers/search_controller.rb:
```diff
@@ -28,8 +45,24 @@ def chat_channel_params
+  end
+
+  def classified_listing_params
+    remove_blank_params
```
Could this be a before_action that applies only to the classified_listings endpoint? Since that’s rails convention, I’d say we should do that unless there’s a specific reason not to!

vaidehijoshi replied:
I’m thinking the action could be something like sanitize_params or the like
